apps/minifront/src/state/swap/price-history.ts

```diff
@@ -1,47 +1,73 @@
+import { sendCandlestickDataRequest } from './helpers';
 
 interface Actions {
-  load: (ac?: AbortController) => AbortController['abort'];
+  /**
+   * History limit becomes the maximum width of the chart domain (block height).
+   */
+  setHistoryLimit: (limit: bigint) => void;
+  /**
+   * Setting history start will cause the chart domain to begin at the specified
+   * block height and extend towards the present. Setting history start to
+   * `undefined` or `0n` will cause the chart domain to end at the present block
+   * height and extend towards the past.
+   */
+  setHistoryStart: (blockHeight?: bigint) => void;
 }
 
+export const { candles, useCandles, useRevalidateCandles } = createZQuery({
+  name: 'candles',
+  fetch: (
+    startMetadata: Metadata | undefined,
+    endMetadata: Metadata | undefined,
+    historyLimit: bigint | undefined,
+    historyStart: bigint | undefined,
+  ) =>
+    Promise.all([
+      sendCandlestickDataRequest(startMetadata, endMetadata, historyLimit, historyStart),
+      sendCandlestickDataRequest(endMetadata, startMetadata, historyLimit, historyStart),
+    ]).then(([direct, inverse]) => ({ direct, inverse })),
+  getUseStore: () => useStore,
+  get: state => state.swap.priceHistory.candles,
+  set: setter => {
+    const newState = setter(useStore.getState().swap.priceHistory.candles);
+    useStore.setState(state => {
+      state.swap.priceHistory.candles = newState;
+    });
+  },
+});
+
 interface State {
-  candles: CandlestickData[];
-  endMetadata?: Metadata;
-  startMetadata?: Metadata;
+  candles: ZQueryState<
+    { direct: CandlestickDataResponse; inverse: CandlestickDataResponse },
+    Parameters<typeof sendCandlestickDataRequest>
+  >;
+  historyLimit: bigint;
+  historyStart?: bigint;
 }
 
 export type PriceHistorySlice = Actions & State;
 
-const INITIAL_STATE: State = {
-  candles: [],
+const INITIAL_STATE: Omit<State, 'pair'> = {
+  candles,
+  historyLimit: PRICE_RELEVANCE_THRESHOLDS.default,
 };
 
-export const createPriceHistorySlice = (): SliceCreator<PriceHistorySlice> => (set, get) => ({
+export const createPriceHistorySlice = (): SliceCreator<PriceHistorySlice> => () => ({
   ...INITIAL_STATE,
-  load: (ac = new AbortController()): AbortController['abort'] => {
-    const { assetIn, assetOut } = get().swap;
-    const startMetadata = getMetadataFromBalancesResponseOptional(assetIn);
-    const endMetadata = assetOut;
-    void sendCandlestickDataRequests(
-      { startMetadata, endMetadata },
-      // there's no UI to set limit yet, and any given range won't always happen
-      // to include price records.
-      PRICE_RELEVANCE_THRESHOLDS.default * 2n,
-      ac.signal,
-    ).then(candles => {
-      set(({ swap }) => {
-        swap.priceHistory.startMetadata = startMetadata;
-        swap.priceHistory.endMetadata = endMetadata;
-        swap.priceHistory.candles = candles;
-      });
+  setHistoryLimit: blocks => {
+    useStore.setState(state => {
```

Comment:
```suggestion
    set(state => {
```

That's the pattern we're using throughout the rest of Zustand (except in ZQuery, which _has_ to use `useStore`).